runtime/caches.go

```diff
@@ -143,21 +156,63 @@ func (c *migrationMetaCache) evict(ctx context.Context, instID string) {
 }
 
 type queryCache struct {
-	cache *lru.Cache
+	cache *ristretto.Cache
+	group *singleflight.Group
 }
 
-func newQueryCache(size int) *queryCache {
-	cache, err := lru.New(size)
+func newQueryCache(sizeInBytes int64) *queryCache {
+	if sizeInBytes <= 0 {
+		panic(fmt.Sprintf("invalid cache size should be greater than 0 : %v", sizeInBytes))
+	}
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		// Use 5% of cache memory for storing counters. Each counter takes roughly 3 bytes.
+		// Recommended value is 10x the number of items in cache when full.
+		// Tune this again based on metrics.
+		NumCounters: int64(float64(sizeInBytes) * 0.05 / 3),
+		MaxCost:     int64(float64(sizeInBytes) * 0.95),
+		BufferItems: 64,
+		Metrics:     true,
+	})
 	if err != nil {
 		panic(err)
 	}
-	return &queryCache{cache: cache}
+
+	observability.Must(meter.RegisterCallback(func(ctx context.Context, observer metric.Observer) error {
+		observer.ObserveInt64(queryCacheHitsCounter, int64(cache.Metrics.Hits()))
+		observer.ObserveInt64(queryCacheMissesCounter, int64(cache.Metrics.Misses()))
+		observer.ObserveInt64(queryCacheItemCountGauge, int64(cache.Metrics.KeysAdded()-cache.Metrics.KeysEvicted()))
+		observer.ObserveInt64(queryCacheSizeBytesGauge, int64(cache.Metrics.CostAdded()-cache.Metrics.CostEvicted()))
```

Comment:
Yes it will gauge the delta but it is the best representation of the cache size based on available metrics. There is no metric for totals : https://github.com/dgraph-io/ristretto/blob/34fd2d33a86ef2eeebb75f3f5d78909b180d7410/cache.go#L507

Reply:
I think we can get the total if you we a counter instead of a gauge, and add `added - evicted` instead of observing it. Would that work?

Why do you think they don't track totals outright? It seems valuable to me – am I missing something?

Reply:
I am just wondering won't the `added-evicted` give the cache size itself and the value of this at any instant would represent the number of items in the cache? 

I am not sure how to use a counter here `added-evicted` is not a delta but the number of added items and evicted items in total so far.

Reply:
Oh, got it – I thought they were deltas for each observation. No problem then